### PR TITLE
Do not allocate shared memory for wal_redo process

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

**Comment by knizhnik** [comment]
shmdt is any case not enabled in secomp mode:
```
		/* Enable those for a proper shutdown.
		PG_SCMP_ALLOW(munmap),
		PG_SCMP_ALLOW(shmctl),
		PG_SCMP_ALLOW(shmdt),
		PG_SCMP_ALLOW(unlink), // shm_unlink
		*/
```
IMHO providing special implementation of PG shared memory (local_shmem.c) just for wal redo is overkill. Also it may be not so convenient to build two different versions of postgres binaries. 

It is also possible to add pecial type for `dynamic_shared_memory_type` (right now sysv and mmap) and look at it. But it is also not so correct, because it referes only dynamic shared memory segments, not static.

From my point of view adding two lines of Neon specific code is better than other alternatives requiring more invasive changes of postgres core.

File: src/backend/port/sysv_shmem.c [hunk 1]
```diff
@@ -155,6 +155,8 @@ InternalIpcMemoryCreate(IpcMemoryKey memKey, Size size)
 		}
 	}
 #endif
+	if (am_wal_redo_postgres)
+		return valloc(size);
```
Comment (hlinnaka) [inline src/backend/port/sysv_shmem.c:159]: Had to look up what `valloc` is, never heard of that before :-).

Why not malloc?

Comment (hlinnaka) [reply src/backend/port/sysv_shmem.c:159]: Please add a comment

**Review by hlinnaka — APPROVED** [review]

**Comment by hlinnaka** [comment]
I agree with Stas' earlier comment that this needs a NULL check. If the valloc() call fails and returns NULL, I think this will retry the valloc() call in an infinite loop.

I also think we should replace the valloc() with plain malloc(). There is no reason for the allocation to be page-aligned.

BTW, with the default settings, this is only used to replace the tiny shmem shim allocation. Most of the shared memory is still allocated with the mmap() call. I think that's OK, as that codepath also does some tricks to make huge pages effective, which probably makes sense for the WAL redo process, too.

**Comment by knizhnik** [comment]
> I agree with Stas' earlier comment that this needs a NULL check. If the valloc() call fails and returns NULL, I think this will retry the valloc() call in an infinite loop.
> 
> I also think we should replace the valloc() with plain malloc(). There is no reason for the allocation to be page-aligned.
> 
> BTW, with the default settings, this is only used to replace the tiny shmem shim allocation. Most of the shared memory is still allocated with the mmap() call. I think that's OK, as that codepath also does some tricks to make huge pages effective, which probably makes sense for the WAL redo process, too.

https://github.com/neondatabase/postgres/pull/173

I do not think that we should eliminate mmaps, because even malloc is using it internally.

**Pull request merged**